visualdl/logic/sdk.h

```diff
@@ -326,6 +326,123 @@ struct TextReader {
   TabletReader reader_;
 };
 
+
+
+
+/*
+ * Image component writer.
+ */
+struct Audio {
+    using value_t = float;
+
+    /*
+     * step_cycle: store every `step_cycle` as a record.
+     * num_samples: how many samples to take in a step.
+     */
+    Audio(Tablet tablet, int num_samples, int step_cycle)
+            : writer_(tablet), num_samples_(num_samples), step_cycle_(step_cycle) {
+      CHECK_GT(step_cycle, 0);
+      CHECK_GT(num_samples, 0);
+
+      writer_.SetType(Tablet::Type::kAudio);
+      // make audio's tag as the default caption.
+      writer_.SetNumSamples(num_samples);
+      SetCaption(tablet.reader().tag());
+    }
+
+    void SetCaption(const std::string& c) {
+      writer_.SetCaptions(std::vector<std::string>({c}));
+    }
+
+    /*
+     * Start a sampling period, this interface will start a new reservior sampling
+     * phase.
+     */
+    void StartSampling();
+    /*
+     * End a sampling period, it will clear all states for reservior sampling.
+     */
+    void FinishSampling();
+
+    /*
+     * A combined interface for IsSampleTaken and SetSample, simpler but might be
+     * low effience.
+     */
+    void AddSample(int sample_rate,
+                   const std::vector<value_t>& data);
+
+    /*
+     * Will this sample be taken, this interface is introduced to reduce the cost
+     * of copy audio data, by testing whether this audio will be sampled, and only
+     * copy data when it should be sampled. In that way, most of unsampled audio
+     * data need not be copied or processed at all.
+     */
+    int IsSampleTaken();
+    /*
+     * Store audio data with sample rate
+     */
+    void SetSample(int index,
+                   int sample_rate,
+                   const std::vector<value_t>& data);
+
+protected:
+    bool ToSampleThisStep() { return step_id_ % step_cycle_ == 0; }
+
+private:
+    Tablet writer_;
+    Record step_;
+    int num_records_{0};
+    int num_samples_{0};
+    int step_id_{0};
+    int step_cycle_;
+};
+
+/*
+* Audio reader.
+*/
+struct AudioReader {
+    using value_t = typename Audio::value_t;
+
+    struct AudioRecord {
+        int step_id;
+        int sample_rate;
+        std::vector<int> data;
```

Comment:
To meet the audio value interval, is `short` or `unsigned char` or `char` is enough? 

https://cn.mathworks.com/help/matlab/ref/audiorecorder.getaudiodata.html?s_tid=gn_loc_drop

here just `int16`, `int8`, `uint8` are used, not `int32`.

Just a suggestion, not that important, `int` works good.

Reply:
you are right, when we were doing speech recognition app before, we just use byte / int8